### PR TITLE
Use scikit-build-core

### PR DESCRIPTION
**Pull request by gmloose**
Use `scikit-build-core` to build the Python casacore bindings. It completely replaces the custom-made stuff in `setup.py`.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+project(python-casacore)
+
+find_package(
+  Python
+  COMPONENTS Interpreter Development.Module
+  REQUIRED)
+
+# Find Casacore and its dependencies
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(Casacore REQUIRED casacore)
```
Comment (tammojan) [inline CMakeLists.txt:11]: This assumes casacore 3.5, which is not yet in ubuntu 22.04. I'd prefer having a fallback based on a FindCasacore.cmake for people who are not on casacore 3.5.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+find_package(Boost REQUIRED COMPONENTS python)
+
```
Comment (tammojan) [inline CMakeLists.txt:13]: CMakeLists should check for the dependencies wcslib and cfitsio.

File: pyproject.toml [hunk 3]
```diff
@@ -1,18 +1,61 @@
...
+    "License :: OSI Approved :: GNU Lesser General Public License v3 (LGPLv3)",
+    "Natural Language :: English",
+    "Operating System :: POSIX :: Linux",
+    "Programming Language :: C++",
+    "Programming Language :: Python :: 3",
+    "Topic :: Scientific/Engineering :: Astronomy",
+]
+dependencies = [
+    "numpy",
+    "six",
```
Comment (tammojan) [inline pyproject.toml:43]: Are those really the only python dependencies?

File: pyproject.toml [hunk 4]
```diff
@@ -1,18 +1,61 @@
...
+# cibuildwheel settings #
+#########################
+
 [tool.cibuildwheel]
 build = "cp3{7,8,9,10,11,12}-*_x86_64"
 build-verbosity = 1
 environment = """ \
-    CFLAGS="-I/usr/include/cfitsio" \
-    LD_LIBRARY_PATH=/opt/boost/lib \
+    CXXFLAGS="-I/usr/include/cfitsio" \
```
Comment (tammojan) [inline pyproject.toml:58]: Should this hardcoded path be removed?

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,38 @@
+python_add_library(_fitting MODULE WITH_SOABI fit.cc fitting.cc)
+
+python_add_library(_functionals MODULE WITH_SOABI functional.cc functionals.cc)
+
+python_add_library(_images MODULE WITH_SOABI images.cc pyimages.cc)
+
+python_add_library(_measures MODULE WITH_SOABI pymeas.cc pymeasures.cc)
+
```
Comment (tammojan) [inline src/CMakeLists.txt:8]: Probably the wcslib and cfitsio linking should happen here, for the _images library.

**Review by tammojan — COMMENTED** [review]
Great to get rid of this setup.py with the awful boost check, and move to CMake and modern python build systems!

File: pyproject.toml (hunk 4 above)
Comment (gmloose) [reply pyproject.toml:58]: TOML does not support variables. But I guess that once a check for `cfitsio` is added to the `CMakeLists.txt` file (as suggested), this is indeed no longer necessary.

File: pyproject.toml (hunk 3 above)
Comment (gmloose) [reply pyproject.toml:43]: These were the only ones listed in `setup.py`. This list seems to be complete. A check revealed there's an additional optional dependency on `wxPython` for the table browser.

Comment (tammojan) [reply pyproject.toml:43]: I was not aware that `wxPython` was a dependency. I don't think it should become a required dependency, as the main uses of python-casacore do not need it and the installation overhead is quite big (like requiring x11). So I'd suggest sticking with `numpy` and `six`.

File: CMakeLists.txt (hunk 1 above)
Comment (gmloose) [reply CMakeLists.txt]: Replaced `pkg-config` stuff with good-ole CMake modules. No need to do fancy stuff :)

File: src/CMakeLists.txt (hunk 5 above)
Comment (gmloose) [reply src/CMakeLists.txt:8]: That happens further down in the `foreach` loop.

**Comment by gmloose** [comment]
I think I addressed all your review comments. Ready for second review.

**Comment by gmloose** [comment]
I do see that my current one-shoe-fits-all solution for linking results in over-linking. I don't think this is a very big issue, because all libraries are in the binary wheel anyway. It will, of course, give a little run-time overhead and a waste of some RAM memory. @tammojan, if you think it's worth the effort to dive into this, I will.

**Comment by gmloose** [comment]
I fixed the over-linking issue in commit [693f7ca](https://github.com/casacore/python-casacore/pull/263/commits/693f7caf188a810612573bb05e20072373b9a15b)

**Comment by tammojan** [comment]
> The OS-X build is again broken. Now with:
> 
> ```
>  CMake Error at /usr/local/Cellar/cmake/3.28.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
>   Could NOT find Python3 (missing: Python3_NumPy_INCLUDE_DIRS NumPy) (found
>   version "3.12.1")
> ```
> 
> I have no idea how to fix this, because I don't know how to reproduce this problem.

It's a problem in the installation of casacore itself. I think this could be solved by making boost-python an optional requirement in casacore's CMake.

File: README.rst [hunk 6]
```diff
@@ -38,30 +42,31 @@ from source
 
 install these requirements:
 
-* `setuptools <https://pypi.python.org/pypi/setuptools>`_
 * `Casacore <https://github.com/casacore/casacore/>`__
 * `Boost-python <http://www.boost.org/libs/python/doc/>`_
-* `numpy <http://www.numpy.org/>`_ 
+* `numpy <http://www.numpy.org/>`_
 * `cfitsio <http://heasarc.gsfc.nasa.gov/fitsio/>`_
+* `wcslib <https://www.atnf.csiro.au/people/mcalabre/WCS/wcslib-8.2.2.tar.bz2>`_
```
[the remark this answers is not in the document]
Comment (gmloose) [reply README.rst]: Done.

**Review by tammojan — APPROVED** [review]
Looks great to me! I think this is ready to merge.

**Pull request merged**